src/System.IO.Pipelines/ReadableBufferReader.cs

```diff
@@ -1,132 +1,178 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void Initalize(BufferSegment segment, int startIndex, int length)
         {
-            get
+            if (length == 0)
             {
-                var part = _enumerator.Current;
-
-                if (_end)
+                _currentSpan = default(Span<byte>);
+            }
+            else
+            {
+                while (segment != null && segment.End == startIndex)
                 {
-                    return new ReadCursor(part.Segment, part.Start + _currentSpan.Length);
+                    segment = segment.Next;
+                    if (segment == null)
+                    {
+                        Debug.Assert(length == 0);
+                        break;
+                    }
+
+                    startIndex = segment.Start;
                 }
+                _currentSpan = segment.Memory.Span.Slice(startIndex, segment.End - startIndex);
+            }
+
+            _cursorOffset = startIndex;
+            _currentSegment = segment;
+            _remainingBytes = length;
+            _consumedBytes = 0;
+            _spanIndex = 0;
+        }
 
-                return new ReadCursor(part.Segment, part.Start + _index);
+        public bool End
+        {
+            get
+            {
+                Debug.Assert(_remainingBytes >= 0);
+                return _remainingBytes == 0;
             }
         }
 
+        public int Index => _spanIndex;
+
+        public ReadCursor Cursor => new ReadCursor(_currentSegment, _cursorOffset + _spanIndex);
+
         public Span<byte> Span => _currentSpan;
 
         public int ConsumedBytes => _consumedBytes;
 
+        public int RemainingBytes => _remainingBytes;
+
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public int Peek()
         {
-            if (_end)
+            if (End)
             {
                 return -1;
             }
-            return _currentSpan[_index];
+            return _currentSpan[_spanIndex];
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public int Take()
         {
-            if (_end)
+            if (End)
             {
                 return -1;
             }
 
-            var value = _currentSpan[_index];
-
-            _index++;
+            _remainingBytes--;
             _consumedBytes++;
+            var value = _currentSpan[_spanIndex];
+            _spanIndex++;
 
-            if (_index >= _currentSpan.Length)
+            if (_spanIndex >= _currentSpan.Length)
             {
-                MoveNext();
+                TraverseSegments(0);
             }
 
             return value;
         }
 
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private void MoveNext()
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Skip(int length)
         {
-            while (_enumerator.MoveNext())
+            if ((uint)_remainingBytes < (uint)length)
```

Comment:
tests both length being greater than remaining and length being negative (as then as `uint` it would be greater than `int.MaxValue`) in single test